freeglut/freeglut/src/mswin/fg_spaceball_mswin.c

```diff
@@ -42,158 +50,563 @@
+        return;
+    }
+
+    if( (pRawInputDeviceList = 
+        malloc(sizeof(RAWINPUTDEVICELIST) * numDevices ) ) == NULL )
+    {
+        fgError( "out of memory - numDevices" );
+        return;
+    }
+
+    if( GetRawInputDeviceList( pRawInputDeviceList, &numDevices, 
+        sizeof(RAWINPUTDEVICELIST) ) == -1 )
+    {
+        fgError( "spaceball list" );
+        free( pRawInputDeviceList );
+        return;
+    }
+
+    for( i=0; i < numDevices; i++ )
+    {
+        RID_DEVICE_INFO devInfo;
+        unsigned int data_size = 1023;
+
+        /*
+            char device_name[1024];
+            printf("dev index: %d type: %d\n",i, pRawInputDeviceList[i].dwType);
+            if (GetRawInputDeviceInfo(
+                    pRawInputDeviceList[i].hDevice,
+                    RIDI_DEVICENAME,
+                    &device_name, &data_size) == -1)
+            {
+                fgError( "HID name" );
+                continue;
+            }
+            printf("    name: %s\n", device_name);
+        */
+
+        devInfo.cbSize = sizeof(RID_DEVICE_INFO);
+        data_size = sizeof(RID_DEVICE_INFO);
+        if( GetRawInputDeviceInfo( pRawInputDeviceList[i].hDevice,
+                RIDI_DEVICEINFO, &devInfo, &data_size ) == -1 ) 
+        {
+            fgError( "spaceball info");
+            continue;
+        }
+
+        if( devInfo.hid.dwVendorId == VENDOR_LOGI
+            || devInfo.hid.dwVendorId == VENDOR_3DX )
+        {
+            int j;
+            
+            /* found a multi-axis device with a known VID */
+            /* set defaults for potential unknown device support */
+            fg_sball_buttons = fg_sball_bitwise_bits = A3DX_DEFAULT_BUTTON_BITS;
+
+            /* set the number of valid bits and buttons based on PID */
+            for( j=0; spaceballList[j].VID > 0; j++ )
+                if( spaceballList[j].PID == devInfo.hid.dwProductId )
+                {
+                    fg_sball_bitwise_bits = spaceballList[j].bits;
+                    fg_sball_buttons = spaceballList[j].buttons;
+                }
+            /*
+                printf( "VID: 0x%04x  PID: 0x%04x  bits: %d  buttons: %d\n",
+                    devInfo.hid.dwVendorId, devInfo.hid.dwProductId, 
+                    fg_sball_bitwise_bits, fg_sball_buttons );
+            */
+        }
+    }
+    free( pRawInputDeviceList );
+}
+
+void fgSpaceballHandleWinEvent( HWND hwnd, WPARAM wParam, LPARAM lParam )
+{
+    SFG_Window* window;
     HRAWINPUT hRawInput = (HRAWINPUT)lParam;
-    UINT inputCode = (UINT)wParam;
-    UINT size;
-    BYTE *rawInputBuffer;
     PRAWINPUT pRawInput;
-    UINT res;
     RID_DEVICE_INFO sRidDeviceInfo;
+    BYTE *rawInputBuffer;
+    UINT size;
 
-    if (!fg_sball_initialized)
+    if( !fg_sball_initialized )
     {
         fgPlatformInitializeSpaceball();
-        if (!fg_sball_initialized)
-        {
-            return;
-        }
+        if( !fg_sball_initialized )
```

Comment:
random formatting change

Reply:
will revert to the original verbose curly brackets.